compiler/src/iree/compiler/Codegen/Dialect/PCF/IR/PCFBase.td

```diff
@@ -0,0 +1,280 @@
+    2. Tokens. Types capable of managing synchronization of resources between
+       threads. This could be anything ranging from fences + (named) barriers to
+       producer/consumer queues implemented with ringbuffers.
+
+    PCF ops + types are designed to be lowered in three phases, starting from
+    structural ops on scoped memory infused with synchronization tokens. Prior
+    to each phase a different level of scheduling is implied.
+
+    1. Tokens tied to resources are split and lowered to separate ops. Before
+       this the compiler can perform coarse grain scheduling around resources
+       according to their tied synchronization.
+    2. Generic scoped memory is converted to memref. Since all tokens have been
+       resolved by this point, this is just a matter of propagating layout and
+       memory space.
+    3. Wrapping structured ops are lowered to control flow (scf and/or cf).
+
+    This diagram illustrates where the dialect fits in to executable
+    lowering pipelines for typical GPUs. For CPUs and other accelerators, the
+    same flow is intended to work modulo different levels of physical
+    parallelism instead of thread/subgroup/lane.
+
+    ```
+            v----------+----------v
+            | Executable Input    |
+            | (Linalg on tensors) |
+            +----------v----------+
+                       |
+     TileAndDistribute |
+     to workgroups     |
+                       |
+            v----------+----------v
+            |    PCF and/or SCF   |
+            +----------v----------+
+                       |
+     SCF(.forall)ToPCF |
+                       |
+             v---------+---------v
+    +--------|-------+   +-------|--------+
+    |   Tile Op1 to  |   |   Tile OpN to  |
+    |   Subgroups/   |   |   Subgroups/   |
+    |   Threads/     |...|   Threads/     |
+    |   Lanes        |   |   Lanes        |
+    | pcf.concurrent |   | pcf.concurrent |
+    +--------|-------+   +-------|--------+
+             +---------v---------+
+                       |
+             Vectorize | // WriteOps vectorize
+                       |
+             Bufferize | // PCF tensor -> pcf.sref
+                       | // becomes memref -> sref
+                       |
+         ResolveTokens |
+          SRefToMemRef | // pcf.sref -> memref
+              LowerPCF | // pcf -> scf/cf
+                       |
+           v-----------+-----------v
+           | SCF+GPU+vector+memref |
+           +-----------------------+
+    ```
+
+  }];
+
+  let dependentDialects = [];
+
+  let useDefaultTypePrinterParser = 1;
+  let useDefaultAttributePrinterParser = 1;
+}
+
+//===----------------------------------------------------------------------===//
+// Base pcf dialect op classes
+//===----------------------------------------------------------------------===//
+
+class PCF_Op<string mnemonic, list<Trait> traits = []> :
+    Op<PCF_Dialect, mnemonic, traits> {
+  let hasCustomAssemblyFormat = 1;
```

Comment:
... Huh

Reply:
good catch. I am shocked this never caused me issues...